service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java

```diff
@@ -0,0 +1,265 @@
+import org.hl7.fhir.dstu3.model.Location;
+import org.hl7.fhir.dstu3.model.Organization;
+import org.hl7.fhir.dstu3.model.ResourceType;
+import org.hl7.fhir.dstu3.model.SimpleQuantity;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.github.mustachejava.Mustache;
+
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+@Component
+public class ImmunizationObservationStatementMapper {
+
+    private static final String UK_ZONE_ID = "Europe/London";
+    private static final Mustache OBSERVATION_STATEMENT_TEMPLATE = TemplateUtils
+        .loadTemplate("ehr_observation_statement_template.mustache");
+    private static final DateTimeFormatter DATE_TIME_FORMATTER_SHORT = new DateTimeFormatterBuilder()
+        .appendPattern("yyyy-MM-dd")
+        .toFormatter();
+
+    private static final String PARENT_PRESENT_URL = "https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-ParentPresent-1";
+    private static final String DATE_RECORDED_URL = "https://fhir.hl7.org.uk/STU3/StructureDefinition/Extension-CareConnect-DateRecorded-1";
+
+    private static final String PARENT_PRESENT = "Parent Present: ";
+    private static final String LOCATION = "Location: ";
+    private static final String MANUFACTURER = "Manufacturer: ";
+    private static final String BATCH = "Batch: ";
+    private static final String EXPIRATION = "Expiration: ";
+    private static final String SITE = "Site: ";
+    private static final String ROUTE = "Route: ";
+    private static final String QUANTITY = "Quantity: ";
+    private static final String REASON = "Reason: ";
+    private static final String REASON_NOT_GIVEN = "Reason not given: ";
+    private static final String VACCINATION_PROTOCOL_STRING = "Vaccination Protocol %S: %S%nSequence: %S,%S%n";
+    private static final String VACCINATION_TARGET_DISEASE = "Target Disease: ";
+    private final MessageContext messageContext;
+
+    public String mapImmunizationToObservationStatement(Immunization immunization, Bundle bundle, boolean isNested) {
+        var observationStatementTemplateParameters = ImmunizationObservationStatementTemplateParameters.builder()
+            .observationStatementId(messageContext.getIdMapper().getOrNew(ResourceType.Immunization, immunization.getId()))
+            .availabilityTime(buildAvailabilityTime(immunization))
+            .effectiveTime(buildEffectiveTime(immunization))
+            .pertinentInformation(buildPertinentInformation(immunization, bundle))
+            .isNested(isNested)
+            .build();
+        return TemplateUtils.fillTemplate(OBSERVATION_STATEMENT_TEMPLATE, observationStatementTemplateParameters);
+    }
+
+    private String buildAvailabilityTime(Immunization immunization) {
+        var dateRecordedExtension = filterExtensionByUrl(immunization, DATE_RECORDED_URL);
+        if (dateRecordedExtension.isPresent()) {
```

Comment:
What do you think about:
`
return formatDateTimeType((DateTimeType) dateRecordedExtension
            .orElseThrow(() -> new EhrMapperException("Could not map recorded date")).getValue());
`